### ISSUE TITLE
How does `as` props work

### ISSUE THREAD

**Issue by TristanWYL** [chakra-ui/chakra-ui#6367]
### Description

I just encountered a problem. When I was using `as` props to transform ChakraUI's `Box` to another 3rd-party functional component, say `Form`, everyting went well, but `Form` has its own `onSubmit` props which collided with that of `Box`. Here was the problem: the type of `onSubmit` is still the same as that of `Box`'s `onSubmit` props, rather than that of `Form`'s. If I want to keep the type of `Form`'s `onSubmit` after using the `as` props of `Box`, what could I do? A [sandbox](https://codesandbox.io/s/epic-christian-3mfvct?file=/src/index.tsx:696-728) was made for reference. Notice the error reported at Line 41.  To try to figure this out, I attempted to find answer by googling or digging into the source code of ChakraUI but in vain. I am very curious about how the `as` props works? Any response will be highly appreciated. Or if you could point out where is the `as` related source code? Thanks so much in advance.

### Link to Reproduction

https://codesandbox.io/s/epic-christian-3mfvct?file=/src/index.tsx:696-728

### Steps to reproduce

2. Click on '...'
3. Scroll down to '...'

### Chakra UI Version

2.2.4

### Operating System

- [ ] macOS
- [ ] Windows
- [ ] Linux

**Comment by itkrt2y** [comment]
Box props are based on `div` tag because `Box` is created by `chakra("div")`.
https://github.com/chakra-ui/chakra-ui/blob/88e812c3a6d99554965b54856382916b6fbf6774/packages/layout/src/box.tsx#L18

If you want to use redwood's `Form` component with chakra style props, I think you should create a new Form component by chakra factory function.
https://chakra-ui.com/docs/styled-system/chakra-factory#chakra-factory-function

```tsx
import { FieldValues, Form as RedwoodForm, useForm } from "@redwoodjs/forms";
import { chakra } from "@chakra-ui/react";

const Form = chakra(RedwoodForm);

function App() {
  const formMethods = useForm<FieldValues>({
    mode: "onTouched",
    defaultValues: { message: "" },
  });

  function handleSubmit(value: { message: string }, e: any) {
    alert(value.message);
    console.log(e);
  }

  return (
    <Form
      onSubmit={handleSubmit}
      formMethods={formMethods}
      borderWidth="1px"
      borderColor="cyan.400"
      borderRadius="2xl"
      w="50%"
      m={8}
      p={4}
    >
      
    </Form>
  );
}
```

**Comment by TristanWYL** [comment]
@itkrt2y Thx! Your reply helps a lot! But I am still very curious how ChakraUI's `as` props works? Would you be able to point out where I could find the source code about this? I want to check out the source code in person if possible.

**Comment by itkrt2y** [comment]
I just read the internals for the first time, so I'm sorry if I made a mistake

## TL;DR

`as` props are processed by emotion, not ChakraUI.
https://emotion.sh/docs/styled#as-prop

## Process

1. Box is created by chakra factory function

https://github.com/chakra-ui/chakra-ui/blob/88e812c3a6d99554965b54856382916b6fbf6774/packages/layout/src/box.tsx#L18

2. The chakra factory function formats the chakra theme into an emotion style object and call emotion with this. In this step, `as` props do nothing.

https://github.com/chakra-ui/chakra-ui/blob/88e812c3a6d99554965b54856382916b6fbf6774/packages/system/src/factory.ts#L16-L23
https://github.com/chakra-ui/chakra-ui/blob/88e812c3a6d99554965b54856382916b6fbf6774/packages/system/src/system.ts#L9
https://github.com/chakra-ui/chakra-ui/blob/88e812c3a6d99554965b54856382916b6fbf6774/packages/system/src/system.ts#L78-L81

3. Default exported function from `"@emotion/styled"` is `createStyled` in emotion.

https://github.com/emotion-js/emotion/blob/b8d45d4587de69a1a6fd80d9543795e3795ca530/packages/styled/src/base.js#L54

This function is `import styled from "@emotion/styled"`'s `styled`.

4. The final tag is decided here

https://github.com/emotion-js/emotion/blob/b8d45d4587de69a1a6fd80d9543795e3795ca530/packages/styled/src/base.js#L107

5. Define component with `FinalTag`

https://github.com/emotion-js/emotion/blob/b8d45d4587de69a1a6fd80d9543795e3795ca530/packages/styled/src/base.js#L168

**Comment by TristanWYL** [comment]
Hi @itkrt2y Thanks so much for your exhaustive explanation, and your reply perfectly solves my concerns. I tried so much to dig into ChakraUI's source code, while ignored that of emotion. Sigh...... How stupid I am. Thanks so much!!!

**Comment by Philzen** [comment]
Thank you @itkrt2y for your very enlightening research here! Using the Chakra factory function is indeed a workaround that works for my use case, which led to my initial report over at https://github.com/redwoodjs/redwood/issues/5956.

Using the factory method, no TS warning is displayed: https://codesandbox.io/s/hungry-http-z8olnf?file=/src/index.tsx ,  whilst using `<chakra.form>` it displays a warning: https://codesandbox.io/s/stupefied-mayer-hvvpwd?file=/src/index.tsx

I always thought using `chakra.<element>` would be equivalent to using the factory function, at least that's the impression i got from the [doc page](https://chakra-ui.com/docs/styled-system/chakra-factory#chakra-jsx-elements) – but apparently they are not equivalent… or should they be?

One thing that still puzzles me: For instance, when one tries to set props on a chakra component that do not exist, one get's a typescript error (which makes sense). As soon as one adds the `as` prop forwarding that prop to an element that actually does have these props, they become available and everything compiles fine. Following the same DX: Shouldn't the prop definition of an `as={…}` component overwrite the prop definition of the Chakra element they are used on? 

I understand from your above research that the `as`-logic is implemented in Emotion – however the `ComponentWithAs` type definition is within the Chakra UI domain, so i'm wondering if there's a way to fix that behaviour.
